### ISSUE TITLE
About the weighted softmax, the forward result is too small

### ISSUE THREAD

**Issue by AmigoCDT** [apache/incubator-mxnet#14229]
I implementate the weighted softmax by `https://github.com/apache/incubator-mxnet/blob/v1.0.0/example/sparse/weighted_softmax_ce.py`

Following is my implementation:
`class WeightedSoftmaxCrossEntropyLoss(mx.operator.CustomOp):
    """
    softmax cross entropy weighted loss, where the loss is adjusted by \
    (class_weight) / sum_of_all_weights)
    """
    def __init__(self, class_weights):
        # parse initial weights from a string to separate items
        self.class_weights = mx.nd.array([float(x) for x in class_weights.split(",")])
        self.class_scales = self.class_weights

    def forward(self, is_train, req, in_data, out_data, aux):
        """Implements forward computation.

        is_train : bool, whether forwarding for training or testing.
        req : list of {'null', 'write', 'inplace', 'add'}, how to assign to out_data. 'null' means skip assignment, etc.
        in_data : list of NDArray, input data.
        out_data : list of NDArray, pre-allocated output buffers.
        aux : list of NDArray, mutable auxiliary states. Usually not used.
        """
        data = in_data[0]
        label = in_data[1]
        pred = mx.nd.SoftmaxOutput(data, label)
        print("pred is ", max(pred[0]))
        self.assign(out_data[0], req[0], pred)

    def backward(self, req, out_grad, in_data, out_data, in_grad, aux):
        """Implements backward computation

        req : list of {'null', 'write', 'inplace', 'add'}, how to assign to in_grad
        out_grad : list of NDArray, gradient w.r.t. output data.
        in_grad : list of NDArray, gradient w.r.t. input data. This is the output buffer.
        """
        label = in_data[1]
        pred = out_data[0]
        print("outgrad[0] is ", out_grad[0][0])
        # move to GPU context if needed
        class_scales = self.class_scales.as_in_context(label.context)
        dx = (pred - mx.nd.one_hot(label, class_scales.shape[0])) * out_grad[0]
        print("dx is ", dx[0])
        scale_factor = (class_scales[label]).reshape((pred.shape[0],1))
        rescaled_dx = scale_factor * dx
        self.assign(in_grad[0], req[0], rescaled_dx)`

But when I use this customop as loss, about 200k IDs, I found that net not be trained, the pred of forward() is all about 2.05e-5 or 4.06e-29, max is about 0.0012, this means my forward() does not work as I supposed.
I dont know where the bug is? And how to solve it?
Please, thank you all so much!

**Comment by frankfliu** [comment]
@mxnet-label-bot add [question, Good first issue]

**Comment by AmigoCDT** [comment]
I found that the sum of pred in forward() is 1, this means the forward() works well.
I will find if the problem happened in backward().

**Comment by piyushghai** [comment]
@AmigoCDT Did you find an issue with the backward pass as well ?

**Comment by AmigoCDT** [comment]
@piyushghai yes, I use another similar customOP (Focal loss), it works well. 
So the trouble is in my code. Two choices : 
 1. gradient:  `gradient passing` or `gradient formulation`.
 2. I lose something in my script. If there is no fault at my gradient. The main error get this:  `mx.metric.EvalMetric`, Focal loss CustomOP also rewrite this class. But I do not know what does this do. I will find at mxnet doc.

**Comment by AmigoCDT** [comment]
@piyushghai I think the error is in backward() or gradient passing. I find few user rewrite `mx.metric.EvalMetric`. This maybe not necessary.
